maven-plugin-tools-annotations/src/site/apt/index.apt

```diff
@@ -94,27 +94,42 @@ public class MyMojo
                 hint = "..." )
     private MyComponent component;
 
+    // pseudo-parameters (marked read-only) permitting injection of Maven build context objects
     // sample objects taken from Maven API through PluginParameterExpressionEvaluator
+    //     https://maven.apache.org/ref/current/maven-core/apidocs/org/apache/maven/plugin/PluginParameterExpressionEvaluator.html
+    // plugins targetting Maven 3.2.5+ (after MNG-5695) should not use these pseudo-parameters any more,
+    // but @Component and Maven APIs to get better compiler-time checks
 
-    @Parameter( defaultValue = "${session}", readonly = true )
+    // @Parameter( defaultValue = "${session}", readonly = true )
+    @Component // since Maven 3.2.5, thanks to MNG-5695
```

Comment:
We have a warning for such components ...

https://github.com/apache/maven-plugin-tools/blob/master/maven-plugin-tools-api/src/main/java/org/apache/maven/tools/plugin/util/PluginUtils.java#L61

and Component is replaced by expression
https://github.com/apache/maven-plugin-tools/blob/master/maven-plugin-tools-annotations/src/main/java/org/apache/maven/tools/plugin/extractor/annotations/JavaAnnotationsMojoDescriptorExtractor.java#L791

Reply:
We were discussing and leaning to the exact opposite.
Keep `@Parameter` for things injected from the config and `@Component/@Inject` for injecting beans, including those from the session (Session) and mojoExecution scopes (MojoExecution, Project)...

Reply:
> We were discussing and leaning to the exact opposite. Keep `@Parameter` for things injected from the config and `@Component/@Inject` for injecting beans, including those from the session (Session) and mojoExecution scopes (MojoExecution, Project)...

This is exactly what I remember. Use `@Parameter` exactly for this case.

Reply:
> > We were discussing and leaning to the exact opposite. Keep `@Parameter` for things injected from the config and `@Component/@Inject` for injecting beans, including those from the session (Session) and mojoExecution scopes (MojoExecution, Project)...
> 
> This is exactly what I remember. Use `@Parameter` exactly for this case.

But those components can also be injected with `@Component`, so this is weird to warn about something that is supported by sisu (because those are actually bound to scopes), while pushing users to a _custom_ solution.  Why only a few components would be accessible by name then ?
I would agree if we had a single annotation for both use cases, but that's not the case in Maven 3.  I suppose we could try for the v4 api though.

Reply:
> > > We were discussing and leaning to the exact opposite. Keep `@Parameter` for things injected from the config and `@Component/@Inject` for injecting beans, including those from the session (Session) and mojoExecution scopes (MojoExecution, Project)...
> > 
> > 
> > This is exactly what I remember. Use `@Parameter` exactly for this case.
> 
> But those components can also be injected with `@Component`, so this is weird to warn about something that is supported by sisu (because those are actually bound to scopes), while pushing users to a _custom_ solution. Why only a few components would be accessible by name then ? I would agree if we had a single annotation for both use cases, but that's not the case in Maven 3. I suppose we could try for the v4 api though.

I agree that this is inconsistent. If both are valid and are functionally identical then the warning is warong.

Reply:
and do we now agree that using `@Component` is better than using `@Parameter( defaultValue = "${session}", readonly = true )` (because smaller and moreover we're sure there is no typo in the default value expression)?

Reply:
> and do we all now agree that using `@Component` is better than using `@Parameter( defaultValue = "${session}", readonly = true )` (because smaller and moreover we're sure there is no typo in the default value expression)?
> 
> if we agree, the warning could be to promote switch to `@Component`...

From my PoV, is this cannot be provided by POM or CLI means and it is unambiguous, then yes.

Reply:
reading git blame, I see that in the past, we (I :) ) implemented in plugin-tools a hack to replace `@Component` with an expression in the plugin descriptor: https://issues.apache.org/jira/browse/MPLUGIN-204

I don't remember when this was done, if it was before Maven 3.2.5 MNG-5695, but with Maven 3.2.5, this hack is not useful for `MavenSession`, `MavenProject` nor `MojoExecution`: it remains useful for a few other types...

Reply:
IMHO there is no problem when JSR330 being used. Problem happens when Mojo APIs (java5 anno or javadoc annos are being used).